### PR TITLE
feat(provider/google): Added cloudrun provider functionality in clouddriver.

### PR DESCRIPTION
**Pull request by sanopsmx**
This is part of https://github.com/spinnaker/governance/issues/302

Added cloudrun provider functionality in clouddriver.

### CODE REVIEW

**Comment by skandragon** [comment]
Can you resolve the conflicts please?

File: clouddriver-cloudrun/src/main/java/com/netflix/spinnaker/clouddriver/cloudrun/cache/Keys.java [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+        break;
+      case CLUSTERS:
+        Names names = Names.parseName(parts[4]);
+        result.put("account", parts[2]);
+        result.put("application", parts[3]);
+        result.put("name", parts[4]);
+        result.put("cluster", parts[4]);
+        // we are not using stack & detail at present
+        //        result.put("stack", names.getStack());
+        //        result.put("detail", names.getDetail());
```
Comment (kirangodishala) [inline clouddriver-cloudrun/src/main/java/com/netflix/spinnaker/clouddriver/cloudrun/cache/Keys.java:87]: are stack & detail not used?

**Comment by kirangodishala** [comment]
@sanopsmx - it seems there are conflicts. Could you resolve them first?

**Comment by sanopsmx** [comment]
> @sanopsmx - it seems there are conflicts. Could you resolve them first?

resolved

**Comment by sanopsmx** [comment]
@kirangodishala @skandragon 

conflicts are resolved.

File: clouddriver-cloudrun/src/main/java/com/netflix/spinnaker/clouddriver/cloudrun/cache/Keys.java (hunk 1 above)
Comment (sanopsmx) [reply clouddriver-cloudrun/src/main/java/com/netflix/spinnaker/clouddriver/cloudrun/cache/Keys.java:87]: Not used.

File: clouddriver-cloudrun/src/main/java/com/netflix/spinnaker/clouddriver/cloudrun/cache/Keys.java [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+        result.put("loadBalancer", parts[3]);
+        break;
+      case SERVER_GROUPS:
+        Names names_1 = Names.parseName(parts[5]);
+        result.put("application", names_1.getApp());
+        result.put("cluster", parts[2]);
+        result.put("account", parts[3]);
+        result.put("region", parts[4]);
+        // we are not using stack & detail at present
+        //        result.put("stack", names_1.getStack());
```
[the remark this answers is not in the document]
Comment (sanopsmx) [reply clouddriver-cloudrun/src/main/java/com/netflix/spinnaker/clouddriver/cloudrun/cache/Keys.java:106]: Not used and deleted.

File: clouddriver-cloudrun/src/main/java/com/netflix/spinnaker/clouddriver/cloudrun/cache/Keys.java [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+        //        result.put("detail", names_1.getDetail());
+        result.put("serverGroup", parts[5]);
+        result.put("name", parts[5]);
+        // we are not using sequence at present
+        //        result.put("sequence", names_1.getSequence().toString());
```
[the remark this answers is not in the document]
Comment (sanopsmx) [reply clouddriver-cloudrun/src/main/java/com/netflix/spinnaker/clouddriver/cloudrun/cache/Keys.java:111]: Deleted

File: clouddriver-cloudrun/src/main/java/com/netflix/spinnaker/clouddriver/cloudrun/config/CloudrunCredentialsConfiguration.java [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+                                : a.getAccountType())
+                        .setProject(a.getProject())
+                        .setJsonKey(jsonKey)
+                        .setApplicationName(clouddriverUserAgentApplicationName)
+                        .setJsonPath(a.getJsonPath())
+                        .setServiceAccountEmail(a.getServiceAccountEmail())
+                        .setLocalRepositoryDirectory(a.getLocalRepositoryDirectory())
+                        .setRequiredGroupMembership(a.getRequiredGroupMembership())
+                        .setPermissions(a.getPermissions().build())
+                        .setSshTrustUnknownHosts(a.isSshTrustUnknownHosts())
```
[the remark this answers is not in the document]
Comment (sanopsmx) [reply clouddriver-cloudrun/src/main/java/com/netflix/spinnaker/clouddriver/cloudrun/config/CloudrunCredentialsConfiguration.java:79]: Not needed.So deleted.

File: clouddriver-cloudrun/src/main/java/com/netflix/spinnaker/clouddriver/cloudrun/config/CloudrunCredentialsConfiguration.java [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+                        .setAccountType(
+                            StringUtils.isEmpty(a.getAccountType())
+                                ? a.getName()
+                                : a.getAccountType())
+                        .setProject(a.getProject())
+                        .setJsonKey(jsonKey)
+                        .setApplicationName(clouddriverUserAgentApplicationName)
+                        .setJsonPath(a.getJsonPath())
+                        .setServiceAccountEmail(a.getServiceAccountEmail())
+                        .setLocalRepositoryDirectory(a.getLocalRepositoryDirectory())
```
[the remark this answers is not in the document]
Comment (sanopsmx) [reply clouddriver-cloudrun/src/main/java/com/netflix/spinnaker/clouddriver/cloudrun/config/CloudrunCredentialsConfiguration.java:76]: Defaulted it to '/tmp' in CloudrunConfigurationProperties.java

File: clouddriver-cloudrun/src/main/java/com/netflix/spinnaker/clouddriver/cloudrun/deploy/ops/DeployCloudrunAtomicOperation.java [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
+   * "freeFormDetails": "details", "credentials": "my-appengine-account", "containerImageUrl":
+   * "gcr.io/my-project/my-image:my-tag", "configFiles": ["env: flex\nruntime:
+   * custom\nmanual_scaling:\n instances: 1\nresources:\n cpu: 1\n memory_gb: 0.5\n disk_size_gb:
+   * 10"] } } ]' "http://localhost:7002/appengine/ops" curl -X POST -H "Content-Type:
+   * application/json" -d '[ { "createServerGroup": { "application": "myapp", "stack": "stack",
+   * "freeFormDetails": "details", "credentials": "my-appengine-credential-name",
+   * "containerImageUrl": "gcr.io/my-gcr-repo/image:tag", "configArtifacts": [{ "type":
+   * "gcs/object", "name": "gs://path/to/app.yaml", "reference": "gs://path/to/app.yaml",
+   * "artifactAccount": "my-gcs-artifact-account-name" }] } } ]'
+   * "http://localhost:7002/appengine/ops"
```
[the remark this answers is not in the document]
Comment (sanopsmx) [reply clouddriver-cloudrun/src/main/java/com/netflix/spinnaker/clouddriver/cloudrun/deploy/ops/DeployCloudrunAtomicOperation.java:214]: Deleted.

File: clouddriver-cloudrun/src/main/java/com/netflix/spinnaker/clouddriver/cloudrun/deploy/ops/DisableCloudrunAtomicOperation.java [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+  @Autowired CloudrunClusterProvider cloudrunClusterProvider;
+
+  public DisableCloudrunAtomicOperation(EnableDisableCloudrunDescription description) {
+    this.description = description;
+  }
+
+  /**
+   * curl -X POST -H "Content-Type: application/json" -d '[ { "disableServerGroup": {
+   * "serverGroupName": "app-stack-detail-v000", "credentials": "my-appengine-account" }} ]'
+   * localhost:7002/appengine/ops
```
[the remark this answers is not in the document]
Comment (sanopsmx) [reply clouddriver-cloudrun/src/main/java/com/netflix/spinnaker/clouddriver/cloudrun/deploy/ops/DisableCloudrunAtomicOperation.java:53]: Deleted

File: clouddriver-cloudrun/src/main/java/com/netflix/spinnaker/clouddriver/cloudrun/op/manifest/CloudrunDeployManifestOperation.java [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+   * "freeFormDetails": "details", "credentials": "my-appengine-account", "containerImageUrl":
+   * "gcr.io/my-project/my-image:my-tag", "configFiles": ["env: flex\nruntime:
+   * custom\nmanual_scaling:\n instances: 1\nresources:\n cpu: 1\n memory_gb: 0.5\n disk_size_gb:
+   * 10"] } } ]' "http://localhost:7002/appengine/ops" curl -X POST -H "Content-Type:
+   * application/json" -d '[ { "createServerGroup": { "application": "myapp", "stack": "stack",
+   * "freeFormDetails": "details", "credentials": "my-appengine-credential-name",
+   * "containerImageUrl": "gcr.io/my-gcr-repo/image:tag", "configArtifacts": [{ "type":
+   * "gcs/object", "name": "gs://path/to/app.yaml", "reference": "gs://path/to/app.yaml",
+   * "artifactAccount": "my-gcs-artifact-account-name" }] } } ]'
+   * "http://localhost:7002/appengine/ops"
```
[the remark this answers is not in the document]
Comment (sanopsmx) [reply clouddriver-cloudrun/src/main/java/com/netflix/spinnaker/clouddriver/cloudrun/op/manifest/CloudrunDeployManifestOperation.java:140]: Deleted.

File: clouddriver-cloudrun/src/main/java/com/netflix/spinnaker/clouddriver/cloudrun/model/CloudrunTrafficSplit.java [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+@Data
+public class CloudrunTrafficSplit implements Cloneable {
+
+  private List<TrafficTarget> trafficTargets = new ArrayList<>();
+
+  @Override
+  public CloudrunTrafficSplit clone() {
+    try {
+      CloudrunTrafficSplit clone = (CloudrunTrafficSplit) super.clone();
+      // TODO: copy mutable state here, so the clone can't change the internals of the original
```
[the remark this answers is not in the document]
Comment (sanopsmx) [reply clouddriver-cloudrun/src/main/java/com/netflix/spinnaker/clouddriver/cloudrun/model/CloudrunTrafficSplit.java:33]: Deleted.

File: clouddriver-cloudrun/src/main/java/com/netflix/spinnaker/clouddriver/cloudrun/deploy/description/UpsertCloudrunLoadBalancerDescription.java [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+    this.loadBalancerName = loadBalancerName;
+  }
+
+  public CloudrunTrafficSplit getSplit() {
+    return split;
+  }
+
+  // public void setSplit(CloudrunTrafficSplit split) {
+  //  this.split = split;
+  // }
```
[the remark this answers is not in the document]
Comment (sanopsmx) [reply clouddriver-cloudrun/src/main/java/com/netflix/spinnaker/clouddriver/cloudrun/deploy/description/UpsertCloudrunLoadBalancerDescription.java:51]: Deleted.

**Review by kirangodishala — COMMENTED** [review]
@sanopsmx - I have given few review comments. Can you fix them?

File: clouddriver-cloudrun/src/main/java/com/netflix/spinnaker/clouddriver/cloudrun/deploy/ops/AbstractStartStopCloudrunAtomicOperation.java [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+  private final StartStopCloudrunDescription description;
+
+  public AbstractStartStopCloudrunAtomicOperation(StartStopCloudrunDescription description) {
+    this.description = description;
+  }
+
+  @Override
+  public Void operate(List priorOutputs) {
+
+    return null;
```
[the remark this answers is not in the document]
Comment (sanopsmx) [reply clouddriver-cloudrun/src/main/java/com/netflix/spinnaker/clouddriver/cloudrun/deploy/ops/AbstractStartStopCloudrunAtomicOperation.java:44]: this operation is not used.

File: clouddriver-cloudrun/src/main/java/com/netflix/spinnaker/clouddriver/cloudrun/description/manifest/CloudrunDeployManifestDescription.java [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+import com.netflix.spinnaker.clouddriver.cloudrun.deploy.description.AbstractCloudrunCredentialsDescription;
+import com.netflix.spinnaker.clouddriver.cloudrun.model.CloudrunService;
+import com.netflix.spinnaker.clouddriver.deploy.DeployDescription;
+import com.netflix.spinnaker.moniker.Moniker;
+import java.util.List;
+import java.util.Map;
+import lombok.Data;
+import lombok.EqualsAndHashCode;
+
+@Data
```
[the remark this answers is not in the document]
Comment (sanopsmx) [reply clouddriver-cloudrun/src/main/java/com/netflix/spinnaker/clouddriver/cloudrun/description/manifest/CloudrunDeployManifestDescription.java:29]: getters and setters removed.

File: clouddriver-cloudrun/src/main/java/com/netflix/spinnaker/clouddriver/cloudrun/security/CloudrunNamedAccountCredentials.java [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+
+  @JsonIgnore private final String localRepositoryDirectory;
+
+  private final CloudrunCredentials credentials;
+
+  private final String applicationName;
+  private final Long cachingIntervalSeconds;
+  private final CloudRun cloudRun;
+
+  @Data
```
[the remark this answers is not in the document]
Comment (sanopsmx) [reply clouddriver-cloudrun/src/main/java/com/netflix/spinnaker/clouddriver/cloudrun/security/CloudrunNamedAccountCredentials.java:55]: lombok.Data removed

File: clouddriver-cloudrun/src/main/java/com/netflix/spinnaker/clouddriver/cloudrun/provider/agent/AbstractCloudrunCachingAgent.java [hunk 14]
```diff
@@ -0,0 +1,102 @@
...
+
+  public AbstractCloudrunCachingAgent(
+      String accountName, ObjectMapper objectMapper, CloudrunNamedAccountCredentials credentials) {
+    this.accountName = accountName;
+    this.objectMapper = objectMapper;
+    this.credentials = credentials;
+  }
+
+  public boolean shouldIgnoreLoadBalancer(String loadBalancerName) {
+    return false; // TODO - need to fix this
```
[the remark this answers is not in the document]
Comment (sanopsmx) [reply clouddriver-cloudrun/src/main/java/com/netflix/spinnaker/clouddriver/cloudrun/provider/agent/AbstractCloudrunCachingAgent.java:35]: We use these methods in CloudrunCachingAgent.java. We need this method.

File: clouddriver-cloudrun/src/main/java/com/netflix/spinnaker/clouddriver/cloudrun/provider/agent/AbstractCloudrunCachingAgent.java [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+  }
+
+  public boolean shouldIgnoreServerGroup(final String serverGroupName) {
+    return false; // TODO - need to fix this
```
[the remark this answers is not in the document]
Comment (sanopsmx) [reply clouddriver-cloudrun/src/main/java/com/netflix/spinnaker/clouddriver/cloudrun/provider/agent/AbstractCloudrunCachingAgent.java:39]: We use these methods in CloudrunCachingAgent.java. We need this method.

File: clouddriver-cloudrun/src/main/java/com/netflix/spinnaker/clouddriver/cloudrun/deploy/ops/AbstractStartStopCloudrunAtomicOperation.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (kirangodishala) [reply clouddriver-cloudrun/src/main/java/com/netflix/spinnaker/clouddriver/cloudrun/deploy/ops/AbstractStartStopCloudrunAtomicOperation.java:44]: I see StopCloudrunAtomicOperation and StartCloudrunAtomicOperation extends this class but this operation is not used. So essentially there is no need of StopCloudrunAtomicOperation & StartCloudrunAtomicOperation. So there is no need of these classes, right?

File: clouddriver-cloudrun/src/main/java/com/netflix/spinnaker/clouddriver/cloudrun/provider/agent/AbstractCloudrunCachingAgent.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (kirangodishala) [reply clouddriver-cloudrun/src/main/java/com/netflix/spinnaker/clouddriver/cloudrun/provider/agent/AbstractCloudrunCachingAgent.java:35]: If CloudrunCachingAgent knows it's always false, why to use it in the first place? If there is a case to ignore caching loadbalancer this makes sense. But looks like it's not. So better remove it from Caching agent as well.

File: clouddriver-cloudrun/src/main/java/com/netflix/spinnaker/clouddriver/cloudrun/provider/agent/AbstractCloudrunCachingAgent.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (kirangodishala) [reply clouddriver-cloudrun/src/main/java/com/netflix/spinnaker/clouddriver/cloudrun/provider/agent/AbstractCloudrunCachingAgent.java:39]: same as above....no need to check in the cachingagent if it's always false

File: clouddriver-cloudrun/src/main/java/com/netflix/spinnaker/clouddriver/cloudrun/security/CloudrunNamedAccountCredentials.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (kirangodishala) [reply clouddriver-cloudrun/src/main/java/com/netflix/spinnaker/clouddriver/cloudrun/security/CloudrunNamedAccountCredentials.java:55]: Please check CloudrunNamedAccountCredentials.Builder class

[the remark this answers is not in the document]
Comment (kirangodishala) [reply clouddriver-cloudrun/src/main/java/com/netflix/spinnaker/clouddriver/cloudrun/security/CloudrunNamedAccountCredentials.java:55]: Please check CloudrunNamedAccountCredentials.Builder class

[the remark this answers is not in the document]
Comment (sanopsmx) [reply clouddriver-cloudrun/src/main/java/com/netflix/spinnaker/clouddriver/cloudrun/security/CloudrunNamedAccountCredentials.java:55]: Yes...there are no getters and setters in CloudrunNamedAccountCredentials.Builder class

File: clouddriver-google/clouddriver-google.gradle [hunk 16]
```diff
@@ -16,7 +16,7 @@ dependencies {
   implementation ("com.google.apis:google-api-services-compute:beta-rev20201102-1.30.10") {
     force = true
   }
-  implementation "com.google.guava:guava"
+  implementation "com.google.guava:guava:31.1-jre"
```
[the remark this answers is not in the document]
Comment (sanopsmx) [reply clouddriver-google/clouddriver-google.gradle:19]: Deleted.

File: clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/snapshot/SaveSnapshotAtomicOperation.groovy [hunk 17]
```diff
@@ -413,11 +413,13 @@ class SaveSnapshotAtomicOperation implements AtomicOperation<Void> {
         case "OPTIMIZE_AVAILABILITY":
           autoscalerMap.autoscaling_policy.cpu_utilization.predictive_method = "optimize_availability"
           break
+        default:
+          autoscalerMap.autoscaling_policy.cpu_utilization.predictive_method = "none"
       }
```
[the remark this answers is not in the document]
Comment (sanopsmx) [reply clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/snapshot/SaveSnapshotAtomicOperation.groovy:418]: Deleted.

File: clouddriver-google/src/main/java/com/netflix/spinnaker/clouddriver/google/provider/agent/GoogleInternalHttpLoadBalancerCachingAgent.java [hunk 18]
```diff
@@ -692,6 +692,10 @@ private static void handleHealthCheck(
       requestPath = healthCheck.getHttpsHealthCheck().getRequestPath();
       hcType = GoogleHealthCheck.HealthCheckType.HTTPS;
     }
+    //    else if (healthCheck.getUdpHealthCheck() != null) {
+    //      port = healthCheck.getUdpHealthCheck().getPort();
+    //      hcType = GoogleHealthCheck.HealthCheckType.UDP;
+    //    }
```
[the remark this answers is not in the document]
Comment (sanopsmx) [reply clouddriver-google/src/main/java/com/netflix/spinnaker/clouddriver/google/provider/agent/GoogleInternalHttpLoadBalancerCachingAgent.java:699]: Deleted.

File: clouddriver-cloudrun/src/main/java/com/netflix/spinnaker/clouddriver/cloudrun/provider/agent/AbstractCloudrunCachingAgent.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (sanopsmx) [reply clouddriver-cloudrun/src/main/java/com/netflix/spinnaker/clouddriver/cloudrun/provider/agent/AbstractCloudrunCachingAgent.java:39]: done.

File: clouddriver-cloudrun/src/main/java/com/netflix/spinnaker/clouddriver/cloudrun/provider/agent/AbstractCloudrunCachingAgent.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (sanopsmx) [reply clouddriver-cloudrun/src/main/java/com/netflix/spinnaker/clouddriver/cloudrun/provider/agent/AbstractCloudrunCachingAgent.java:35]: done.

File: clouddriver-cloudrun/src/main/java/com/netflix/spinnaker/clouddriver/cloudrun/deploy/ops/AbstractStartStopCloudrunAtomicOperation.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (sanopsmx) [reply clouddriver-cloudrun/src/main/java/com/netflix/spinnaker/clouddriver/cloudrun/deploy/ops/AbstractStartStopCloudrunAtomicOperation.java:44]: Done..deleted.

File: clouddriver-google/src/test/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/SerializeApplicationAtomicOperationUnitSpec.groovy [hunk 19]
```diff
@@ -225,7 +225,6 @@ class SerializeApplicationAtomicOperationUnitSpec extends Specification {
     then:
       resourceMap.google_compute_instance_group_manager[SERVER_GROUP_NAME] == serverGroupMap
       resourceMap.google_compute_instance_template[INSTANCE_TEMPLATE_NAME] == instanceTemplateMap
-      resourceMap.google_compute_autoscaler[SERVER_GROUP_NAME] == autoscalingMap
```
[the remark this answers is not in the document]
Comment (sanopsmx) [reply clouddriver-google/src/test/groovy/com/netflix/spinnaker/clouddriver/google/deploy/ops/SerializeApplicationAtomicOperationUnitSpec.groovy:228]: deleted.

**Review by kirangodishala — APPROVED** [review]
looks good. I approve.

**Comment by j-sandy** [comment]
Checked build, testcases passed and deployed image for runtime verification. Service started without issue. Changes looks good to me.

**Pull request merged**